\chapter{}
\section{Substitutional matrix}
How to quantify the amino acids similarity? This is the question that the biologists set themselves when they started thinking about substitutional matrix.
The substitutional matrices are based on the probability that an AA is substituted with another during evolution. These matrices assign to each pair of AAs, a value that indicates the degree of similarity. They are obtained with statistical methods that reﬂect the frequency with which each AA is replaced by another in homologous proteins families [13]. A positive value in the matrix (Fig. 2.6) means that the two amino acids are similar and they are frequently exchanged each other. \\
The most common used substitutional matrices are PAM and Blosum.
\subsection{BLOSUM matrix}
The BLOSUM matrices were introduced in 1992 by Henikoﬀ & Henikoﬀ [15] to give a score for substitution in the amino acid sequence comparisons.
BLOSUM (BLOck SUbstitution Matrix) matrices are constructed using multiple alignments of evolutionarily divergent proteins. The probabilities used in the matrix calculation are computed by looking at ”blocks” of conserved sequences found in multiple protein alignments. Each block normally refers to a set of proteins in relation to each other. When the frequency of replacement within a family is determined, is possible to individuate the signiﬁcant substitutions. By means of aggregation techniques, all sequences contained in a block are put together into the groups. The numerical value associated to the matrices (eg BLOSUM62) represents the threshold value applied by the method of aggregation. A value of 62 indicates that the sequences are put together in the same group if they have an identity value equal to or greater than 62%. Lower values of threshold (eg BLOSUM45) indicate that they have been grouped sequences that are more distant evolutionarily.

\section{Smith-Waterman local alignment method}
2.3.2.1 Smith-Waterman
Smith and Waterman in 1981 described a Local alignment method [24] which was a variation of N-W algorithm. Aim of this method is to ﬁnd common regions between two protein (Qry, Sbj) through calculation of similarity score. The local methods usage are indicated when long biological sequences are compared. Since in these cases, Global alignment can lead to weak correlation when comparing two highly diverged sequences where only part of them are aligned [25]. S-W, as N-W, is a dynamic programming algorithm (Fig. 2.13) that works with a score matrix (N+1 * M+1) where ﬁrst row and column are initialized to zeros. This algorithm can be substantially described in three steps:
